### DISCUSSION TITLE
Accessing assigned port when port 0 is used

### DISCUSSION THREAD

**Discussion by zelakus** [FirstGearGames/FishNet#430] [Feature Request]
As of FishNet Pro 3.10.4, using Tugboat transport there is no way to access the assigned port when using 0 to make use of any available port. This is a much needed feature as I am launching as many server instances as necessary on a dedicated server and would like to avoid any kind of clashing or race conditions while picking the port across processes. And while it works, FishNet isn't providing the assigned port.

The way server was started is as such:
```C#
TugboatInstance.SetServerBindAddress("0.0.0.0", FishNet.Transporting.IPAddressType.IPv4);
TugboatInstance.SetServerBindAddress("0:0:0:0:0:0:0:0", FishNet.Transporting.IPAddressType.IPv6);
TugboatInstance.SetPort(0);
TugboatInstance.StartConnection(server: true);
```

Behavior was as expected, it used an available port and started successfully however trying to get the port with the following did not work out:
```C#
TugboatInstance.OnServerConnectionState += OnServerConnectionState;
//...
void OnServerConnectionState(FishNet.Transporting.ServerConnectionStateArgs args)
    {
        switch (args.ConnectionState)
        {
            case FishNet.Transporting.LocalConnectionState.Started:
                Debug.Log($"Listening on {TugboatInstance.GetServerBindAddress(FishNet.Transporting.IPAddressType.IPv4)}:{TugboatInstance.GetPort()}");
                break;
        }
    }
```
Result was `0.0.0.0:0` giving me the port I've provided earlier, not what the server's socket is bind to.

As LiteNetLib already provides the said port with the LocalPort property on it's NetManager, this is also accessible in the `FishNet.Transporting.Tugboat.Server.ServerSocket::ThreadedSocket` method. Perhaps `_server.LocalPort` can be set as Tugboat's `_port` upon success while enqueueing the `LocalConnectionState.Started`.

This would make the underlying feature actually usable as the assigned port would be known.

**Note:** This port 0 feature isn't something special to LiteNetLib, it's just what I have as my transport currently.

**Comment by FirstGearGames** [comment]
Am I understanding that you are listening on port 0 which is automatically assigning a port within LiteNetLib, but you cannot fetch that port at runtime as it only returns 0?

If so, I could make the transport try to fetch from the active port if started otherwise provide the configured port.
This may not work for all transports but for LNL it certainly seems possible.

**Reply by zelakus** [reply]
Yes, that's the situation I'm having right now.

I think that's a plausible solution and would be much appreciated!

**Reply by FirstGearGames** [reply]
Was adding this in and noticed LNL returns an int for the port ... that's annoying but suppose I could add in some additional overflow checks.

**Comment by FirstGearGames** [comment]
Try this please
[tugboat_getport.zip](https://github.com/FirstGearGames/FishNet/files/13256983/tugboat_getport.zip)

**Reply by zelakus** [reply]
Using FishNet Pro 3.11.5hf0
Tested on:
- Unity Editor 2022.3.5f1 (Windows x64)
- Dedicated Server Build (StandaloneWindows)
- Dedicated Server Build (Linux64)

Seems to work on all, great work!

**Comment by FirstGearGames** [comment]
This was added.
